### ISSUE TITLE
models not reproducible on cpu

### ISSUE THREAD

**Issue by rebidaldal** [microsoft/LightGBM#2598]
## Environment info

Operating System:  Ubuntu 18.04.3 LTS

CPU/GPU model: Xeon E-5 2686 v4 * 2

C++/Python/R version: Python 3.7.4

LightGBM version or commit hash: 2.3.1

## Error message

Using the same dataset and seed with lightgbm python package on cpu, I am unable to reproduce results I have attained previously on my experiments.  I have double checked the data and parameters from model file dumps and they are matching exactly. In these files I see that initial 24 trees are exactly same. However some tiny divergence occurs in Tree 24 and it subsequently builds up to entirely different trees around Tree 1000. The final results show significant differences with noticeable variations on performance. 

## Reproducible examples

Parameters : 

 "nround"                 : 1600,
"bagging_freq"       : 8,
"bagging_fraction"   : 0.6,    
"feature_fraction"   : 0.65,
"feature_fraction_bynode"   : 0.2,
"lambda_l1"          : 5.77641,
"lambda_l2"          : 2.33872,    
"learning_rate"      : 0.05, 
"max_bin"            : 511,
"max_delta_step"     : 445.15,
"max_depth"          : 9,
"min_data_in_bin"    : 30111, 
"min_data_in_leaf"   : 10340,
"min_gain_to_split"  : 1.0398,
"min_sum_hessian_in_leaf"   : 0.852574,
"num_leaves"         : 255,
"num_threads"        : 15,
"boosting_type"      : "gbrt",
"metric"             : "rmse",
"objective"          : "regression_l2",

Data :  5507488 samples with 660 features

Model files are attached below (feature names are removed)

[issue.zip](https://github.com/microsoft/LightGBM/files/3896785/issue.zip)

**Comment by guolinke** [comment]
we cannot guarantee the reproducible over different code versions.

**Comment by rebidaldal** [comment]
The results are not reproducible when running the same code, with same lightgbm/python etc versions.

**Comment by rebidaldal** [comment]
In fact I have tried running 5 instances of this experiment using same AMI on AWS today and got 5 different results.

**Comment by guolinke** [comment]
could you provide the reproduce code?

**Comment by rebidaldal** [comment]
Since I am using a private dataset I cannot share the exact. I will try to create another example using public data and share it here. The next best thing I can share now are the model files attached in original submission.

**Comment by guolinke** [comment]
I just check your model file, actually most of them are almost the same. 
I guess the problem is caused by the floating sum up error. you can try the `num_threads=1`

**Comment by rebidaldal** [comment]
results are stable with `num_threads=1` however it is extremely slow with reasonably large data. 
Can you tell me which parameter is causing this floating point sum error in threaded settings?

I will try to disable bagging, feature sampling by tree and feature sampling by node to make it reproducible on multiple threads again. My assumption was that all of the random samplings are thread safe and reproducible when training on cpu. (with fixed seed, bagging_seed and feature_fraction_seed)

[corr10.xlsx](https://github.com/microsoft/LightGBM/files/3901880/corr10.xlsx)

**Comment by guolinke** [comment]
There are many floating point sum up by multi-threading, you can check the code with omp and reduction, and remove the omp over these fors.

**Comment by rebidaldal** [comment]
After more experiments I have concluded that bagging is not thread safe at all. (Even with fixed number of threads). Attached below is the kendall tau correlation of model predictions. With all other parameters active, the only way to get perfect stability and avoid floating point sum up errors is to disable bagging.  

[corr12 (2).xlsx](https://github.com/microsoft/LightGBM/files/3905378/corr12.2.xlsx)
